waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,169 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/options,
+  stew/results,
+  chronicles,
+  chronos,
+  metrics,
+  libp2p/protocols/protocol,
+  libp2p/stream/connection,
+  libp2p/crypto/crypto,
+  eth/p2p/discoveryv5/enr
+import
+  ../common/nimchronos,
+  ../common/enr,
+  ../waku_core,
+  ../waku_enr,
+  ../node/peer_manager/peer_manager
+
+logScope:
+  topics = "waku sync"
+
+const WakuSyncCodec* = "/vac/waku/sync/1.0.0"
+const DefaultFrameSize = 153600 # using a random number for now
+const DefaultSyncInterval = 60.minutes
+
+type
+  WakuSyncCallback* = proc(hashes: seq[WakuMessageHash]) {.closure, gcsafe, raises: [].}
+
+  WakuSync* = ref object of LPProtocol
+    peerManager: PeerManager
+    maxFrameSize: int # Not sure if this should be protocol defined or not...
+    syncInterval: Duration
+    callback: Option[WakuSyncCallback]
+
+proc ingessMessage*(self: WakuSync, pubsubTopic: PubsubTopic, msg: WakuMessage) {.async.} =
+  if msg.ephemeral:
+    return
+
+  let msgHash = computeMessageHash(pubsubTopic, msg)
+
+  #TODO call bindings to store the new msg hash
+
+proc serverReconciliation(self: WakuSync, message: seq[byte]): Future[Result[seq[byte], string]] {.async.} =
+  #TODO call binding to compute the payload
+
+  let payload: seq[byte] = @[0]
+
+  ok(payload)
+
+proc clientReconciliation(self: WakuSync, message: seq[byte], hashes: var seq[WakuMessageHash]): Future[Result[Option[seq[byte]], string]] {.async.} =
+  #TODO call binding to compute the payload
+
+  let payload: seq[byte] = @[0]
+
+  # TODO update the hashes
+
+  ok(some(payload))
+
+proc intitialization(self: WakuSync): Future[Result[seq[byte], string]] {.async.} =
+  #TODO call binding to compute the payload
+
+  let payload: seq[byte] = @[0]
+
+  ok(payload)
+
+proc request(self: WakuSync, conn: Connection): Future[Result[seq[WakuMessageHash], string]] {.async, gcsafe.} =
+  let request = (await self.intitialization()).valueOr:
+    return err(error)
+
+  let writeRes = catch: await conn.writeLP(request)
```

Comment:
Presumably to be encapsulated within a protobuf at some point?

Reply:
> Presumably to be encapsulated within a protobuf at some point?

Maybe but why? What would be the benefit? One case could be if we want to negotiate `maxFrameSize` between peers. I'm not sure yet.